_posts/2018-09-15-zkSNARK-in-Ethereum.md

```diff
@@ -0,0 +1,61 @@
+---
+layout:     post
+title:      zkSNARK in Ethereum
+date:       2018-09-15
+summary:    Zero-knowledge proofs in Ethereum
+categories: ethereum
+---
+
+![crypto-lock](https://i.imgur.com/KQEORrb.jpg)
+
+### Background
+
+This week we, at Mana Project, finished implementing Byzantium hard fork's EIPs (The Ethereum Improvement Proposals). The hardest EIPs to implement were EIPs related to zkSNARKs - [EIP-196](https://eips.ethereum.org/EIPS/eip-196) ([implementation in Mana](https://github.com/poanetwork/mana/pull/397)), [EIP-197](https://eips.ethereum.org/EIPS/eip-197)  ([implementation in Mana](https://github.com/poanetwork/mana/pull/406)).
+
+To mark this milestone, In this post, I'll try to give a short summary of a zero-knowledge proof and try to describe why it can be useful for Ethereum without going into too much detail because it uses pretty advanced elliptic curve cryptography based on elliptic curves over finite fields. But if you're interested in implementation details you can visit [BN](https://github.com/poanetwork/bn) - our implementation of BN128 elliptic curve operations.
+
+### Zero-knowledge proof
+
+Zero-Knowledge Succinct Non-interactive ARgument of Knowledge, or zkSNARKs, is a form of zero-knowledge proof. A zero-knowledge proof is a type of cryptography that allows one party (the prover) to prove to another party (the verifier) that it possesses some information without revealing any details about information itself.
+
+Zero-knowledge proof must have three properties:
+- Completeness. If the statement is true, the verifier will be convinced that the prover is honest.
+- Soundness If the statement is false, nothing can convince the verifier that the statement is true.
+- Zero-knowledge. If the statement is true, the verifier can not learn any additional information from the statement.
+
+#### Example
+
+Let's see the easiest and the most understandable example of a zero-knowledge proof that was inspired by the work of Professor Oded Goldreich.
```

Comment:
The following example is inspired by the work of Professor Oded Goldeich. In this scenario, you are the prover and your friend is the verifier.

Imagine your friend is colorblind (and you are not). He has a red ball and a green ball, one in each hand. You task is to prove to him the 2 balls are different colors without giving him any additional information. He takes the balls behind his back and either switches them between hands or keeps them in the same hand. He brings them back out and you inform him whether or not he switched. 

Because you can see the colors, you can tell him each time with certainty whether or not he switched.  After a number of rounds, your friend will become convinced the balls are different colors because you accurately tell him whether or not he switched each time. Your statements are always true, and your friend gains no knowledge about which ball is red and which is green.